### PR TITLE
reading from the command line for ./manager and ./server

### PR DESCRIPTION
**Pull request by dinoallo**
Since sealfs already depends on `clap`, the command line support is implemented using this library.
This commit separates Args and Properties structures. This way, it's more convenient to add more arguments in the future without restructuring Properties.

### CODE REVIEW

**Comment by uran0sH** [comment]
remove 'feature' from commit's title

File: src/bin/manager.rs [hunk 1]
```diff
@@ -19,9 +32,33 @@ async fn main() -> anyhow::Result<()> {
...
     builder.init();
-    //read from yaml.
-    let yaml_str = include_str!("../../examples/manager.yaml");
-    let properties: Properties = serde_yaml::from_str(yaml_str).expect("manager.yaml read failed!");
+    let mut properties: Properties;
+    // read from command line.
+    let args = Args::parse();
+    // if the user provides the config file, parse it and use the arguments from the config file.
+    match args.config_file {
+        None => {
```
Comment (uran0sH) [inline src/bin/manager.rs:40]: If none, we should parse through command line

File: src/bin/manager.rs [hunk 2]
```diff
@@ -19,9 +32,33 @@ async fn main() -> anyhow::Result<()> {
...
         .filter(None, log::LevelFilter::Debug);
     builder.init();
-    //read from yaml.
-    let yaml_str = include_str!("../../examples/manager.yaml");
-    let properties: Properties = serde_yaml::from_str(yaml_str).expect("manager.yaml read failed!");
+    let mut properties: Properties;
+    // read from command line.
+    let args = Args::parse();
+    // if the user provides the config file, parse it and use the arguments from the config file.
+    match args.config_file {
```
Comment (uran0sH) [inline src/bin/manager.rs:39]: let properties = match args.config_file {}

**Review by uran0sH — CHANGES_REQUESTED** [review]

File: src/bin/server.rs [hunk 3]
```diff
@@ -33,12 +56,59 @@ async fn main() -> anyhow::Result<(), Box<dyn std::error::Error>> {
         .filter(None, log::LevelFilter::Debug);
     builder.init();
 
-    //todo
-    //read from command line.
+    let mut properties: Properties;
+    // read from command line.
+    let args = Args::parse();
+    // if the user provides the config file, parse it and use the arguments from the config file.
+    match args.config_file {
```
Comment (uran0sH) [inline src/bin/server.rs:63]: the same above

File: src/bin/server.rs [hunk 4]
```diff
@@ -33,12 +56,59 @@ async fn main() -> anyhow::Result<(), Box<dyn std::error::Error>> {
...
+    properties.storage_path = if let Some(p) = args.storage_path {
+        p
+    } else {
+        properties.storage_path
+    };
+    properties.heartbeat = if let Some(h) = args.heartbeat {
+        h
+    } else {
+        properties.heartbeat
+    };
```
Comment (uran0sH) [inline src/bin/server.rs:110]: a little complicated

**Review by uran0sH — CHANGES_REQUESTED** [review]

**Comment by uran0sH** [comment]
Add an argument for determining whether using the command line or config file

**Comment by dinoallo** [comment]
All are resolved by 8a02af8d38510e9517ce4bc427637ad487ff5f4f

File: src/bin/server.rs [hunk 5]
```diff
@@ -33,12 +60,43 @@ async fn main() -> anyhow::Result<(), Box<dyn std::error::Error>> {
         .filter(None, log::LevelFilter::Debug);
     builder.init();
 
-    //todo
-    //read from command line.
+    // read from default configuration.
+    let default_yaml_str = include_str!("../../examples/server.yaml");
```
Comment (uran0sH) [inline src/bin/server.rs:64]: I think "/etc/sealfs/server.yaml" might be better

**Review by uran0sH — COMMENTED** [review]
fix the ci

File: src/bin/server.rs (hunk 5 above)
Comment (dinoallo) [reply src/bin/server.rs:64]: Definitely. The reason why it's not modified is that testing will be more convenient. This will be fixed in the next commit.

Comment (luanshaotong) [reply src/bin/server.rs:64]: I think it's alright.

**Review by uran0sH — APPROVED** [review]

**Pull request merged**